Builder exits without committing changes, wasting entire build cycle

## Problem

The builder phase can modify files in the worktree and then exit without committing, pushing, or creating a PR. The shepherd's post-builder validation catches this (`commits_ahead=0, uncommitted=N files`) but cannot recover — the entire build cycle is wasted.

**Observed scenario:**
```
[ERROR] builder phase validation failed: worktree exists 
(branch=feature/issue-2546, commits_ahead=0, uncommitted=3 files); 
remote branch missing; no PR; labels=[loom:issue, loom:curated]
```

The builder made meaningful changes to 3 files (cli.py, builder.py, test_cli.py) with correct implementations and passing tests, but exited before running `git add/commit/push/gh pr create`.

## Expected Behavior

When the builder exits and validation detects uncommitted changes with `commits_ahead=0`:

**Option A (recovery):** The shepherd should attempt to commit/push/create-PR on behalf of the builder, since the changes exist and tests passed.

**Option B (retry with context):** The shepherd should retry the builder with context about the uncommitted changes, instructing it to commit and create a PR rather than re-implementing from scratch.

**Option C (better builder instructions):** Ensure the builder always commits before exiting, even if it's unsure about the implementation quality. A commit with changes is recoverable; an exit without commit is not.

## Impact

- Wastes 5-7 minutes of builder time
- Requires human intervention or a full retry cycle
- The work is technically complete but lost due to missing commit

--- a/defaults/.claude/commands/builder.md
+++ b/defaults/.claude/commands/builder.md
@@ -242,7 +242,7 @@ Before writing any code, confirm ALL of these:
 
 ## Progress Checkpoints
 
-**IMPORTANT**: Write checkpoints as you progress through implementation stages. Checkpoints allow the shepherd to detect partial progress if you fail, enabling smarter recovery instead of always retrying from scratch.
+**CRITICAL: Write checkpoints at every stage to enable recovery.** Without checkpoints, the shepherd cannot reliably distinguish "builder made real progress but crashed" from "builder never started meaningful work." While the shepherd can now detect some cases of uncommitted work via log analysis and file counts, checkpoints remain the primary and most reliable signal for recovery. Always write them — skipping checkpoints risks your completed work being retried from scratch instead of recovered.
 
 ### Checkpoint Stages
 

--- a/loom-tools/src/loom_tools/shepherd/phases/builder.py
+++ b/loom-tools/src/loom_tools/shepherd/phases/builder.py
@@ -3107,14 +3107,11 @@ class BuilderPhase:
         Returns True if the worktree exists and any of these conditions hold:
         - Has commits ahead of main (needs: push, PR)
         - Has uncommitted changes AND a checkpoint exists (builder made progress)
+        - Has uncommitted changes with evidence of implementation activity
+          (log shows Edit/Write tool calls) or multiple substantive files
         - Remote branch exists but no PR (needs: PR creation)
         - PR exists but missing loom:review-requested label (needs: label)
         - Checkpoint indicates recoverable stage (tested, committed, pushed)
-
-        When uncommitted changes exist but no checkpoint is present and
-        commits_ahead == 0, returns False — the builder never started
-        meaningful work, so a full retry is more appropriate than a
-        completion phase.
         """
         if not diag.get("worktree_exists"):
             return False
@@ -3127,14 +3124,21 @@ class BuilderPhase:
 
         if diag.get("has_uncommitted_changes", False):
             # Uncommitted changes exist but no commits ahead.
-            # Only treat as incomplete if a checkpoint proves the builder
-            # actually started meaningful work.  Without a checkpoint,
-            # the builder never progressed past initialisation and a
-            # full retry is more appropriate than a completion phase.
+            # Treat as incomplete if any of these prove meaningful work:
+            # 1. A checkpoint exists (builder explicitly reported progress)
             checkpoint_stage = diag.get("checkpoint_stage")
             if checkpoint_stage is not None:
                 return True
-            # No checkpoint → builder never started, not incomplete
+            # 2. Builder log shows implementation activity (Edit/Write tool calls)
+            #    — the builder was actively coding, not just initialising
+            if diag.get("log_has_implementation_activity", False):
+                return True
+            # 3. Multiple substantive uncommitted files suggest real work
+            #    (a single file could be a marker or artifact, but ≥2 is
+            #    strong evidence of implementation)
+            if diag.get("uncommitted_file_count", 0) >= 2:
+                return True
+            # No evidence of meaningful work → full retry is more appropriate
             return False
 
         # Remote branch exists but no PR — only incomplete if there are

--- a/loom-tools/src/loom_tools/validate_phase.py
+++ b/loom-tools/src/loom_tools/validate_phase.py
@@ -756,20 +756,126 @@ def validate_builder(
                 "No substantive changes to recover (only marker files found).",
             )
 
-    # No auto-recovery: fail with diagnostics
-    # Builder failed to create a PR. Mark the issue with failure label and
-    # provide diagnostics for manual intervention.
-    diag = _gather_builder_diagnostics(issue, worktree, repo_root)
-    _mark_phase_failed(
-        issue, "builder",
-        "Builder did not create a PR. Worktree has uncommitted or unpushed changes "
-        "that require manual review.",
-        repo_root, diag.to_markdown(),
-        failure_label="loom:failed:builder",
+    # Attempt mechanical recovery: stage, commit, push, create PR.
+    # The builder produced substantive changes but exited before completing
+    # the git/PR workflow.  We can finish the mechanical steps directly.
+    branch = f"feature/issue-{issue}"
+
+    # Step 1: Stage and commit if there are uncommitted changes
+    if status_output:
+        # Extract meaningful file paths from porcelain output
+        files_to_stage = []
+        for line in substantive:
+            path = line[3:].strip().strip('"') if len(line) > 3 else line.strip()
+            if path:
+                files_to_stage.append(path)
+
+        if files_to_stage:
+            r = subprocess.run(
+                ["git", "-C", worktree, "add", "--"] + files_to_stage,
+                capture_output=True, text=True, check=False,
+            )
+            if r.returncode != 0:
+                diag = _gather_builder_diagnostics(issue, worktree, repo_root)
+                _mark_phase_failed(
+                    issue, "builder",
+                    f"Recovery failed: git add failed: {r.stderr.strip()[:200]}",
+                    repo_root, diag.to_markdown(),
+                    failure_label="loom:failed:builder",
+                )
+                return ValidationResult(
+                    "builder", issue, ValidationStatus.FAILED,
+                    "Recovery failed: could not stage changes.",
+                )
+
+            commit_msg = f"feat: implement changes for issue #{issue}"
+            r = subprocess.run(
+                ["git", "-C", worktree, "commit", "-m", commit_msg],
+                capture_output=True, text=True, check=False,
+            )
+            if r.returncode != 0:
+                diag = _gather_builder_diagnostics(issue, worktree, repo_root)
+                _mark_phase_failed(
+                    issue, "builder",
+                    f"Recovery failed: git commit failed: {r.stderr.strip()[:200]}",
+                    repo_root, diag.to_markdown(),
+                    failure_label="loom:failed:builder",
+                )
+                return ValidationResult(
+                    "builder", issue, ValidationStatus.FAILED,
+                    "Recovery failed: could not commit changes.",
+                )
+
+    # Step 2: Push the branch
+    r = subprocess.run(
+        ["git", "-C", worktree, "push", "-u", "origin", branch],
+        capture_output=True, text=True, check=False,
+    )
+    if r.returncode != 0:
+        diag = _gather_builder_diagnostics(issue, worktree, repo_root)
+        _mark_phase_failed(
+            issue, "builder",
+            f"Recovery failed: git push failed: {r.stderr.strip()[:200]}",
+            repo_root, diag.to_markdown(),
+            failure_label="loom:failed:builder",
+        )
+        return ValidationResult(
+            "builder", issue, ValidationStatus.FAILED,
+            "Recovery failed: could not push branch.",
+        )
+
+    # Step 3: Create PR
+    # Fetch issue title for the PR title
+    r_title = _run_gh(
+        ["issue", "view", str(issue), "--json", "title", "--jq", ".title"],
+        repo_root,
+    )
+    pr_title = r_title.stdout.strip() if r_title.returncode == 0 and r_title.stdout.strip() else f"Issue #{issue}"
+
+    r = subprocess.run(
+        [
+            "gh", "pr", "create",
+            "--head", branch,
+            "--title", pr_title,
+            "--label", "loom:review-requested",
+            "--body", f"Closes #{issue}",
+        ],
+        cwd=repo_root,
+        capture_output=True, text=True, check=False,
+    )
+    if r.returncode != 0:
+        diag = _gather_builder_diagnostics(issue, worktree, repo_root)
+        _mark_phase_failed(
+            issue, "builder",
+            f"Recovery failed: gh pr create failed: {r.stderr.strip()[:200]}",
+            repo_root, diag.to_markdown(),
+            failure_label="loom:failed:builder",
+        )
+        return ValidationResult(
+            "builder", issue, ValidationStatus.FAILED,
+            "Recovery failed: could not create PR.",
+        )
+
+    # Extract PR number from output (format: "https://github.com/.../pull/123")
+    pr_url = r.stdout.strip()
+    recovered_pr = _parse_pr_number(pr_url.split("/")[-1] if "/" in pr_url else pr_url)
+
+    _report_milestone(
+        "heartbeat", task_id, repo_root,
+        action=f"recovery: created PR from uncommitted worktree changes for issue #{issue}",
+    )
+    _log_recovery_event(
+        issue=issue,
+        recovery_type="commit_and_pr",
+        reason="validation_failed",
+        repo_root=repo_root,
+        worktree_had_changes=bool(status_output),
+        pr_number=recovered_pr,
     )
     return ValidationResult(
-        "builder", issue, ValidationStatus.FAILED,
-        "No PR found. Builder failed without creating PR (no auto-recovery).",
+        "builder", issue, ValidationStatus.RECOVERED,
+        f"Recovered: staged, committed, pushed, and created PR from worktree changes",
+        "commit_and_pr",
     )
 
 

--- a/loom-tools/tests/shepherd/test_phases.py
+++ b/loom-tools/tests/shepherd/test_phases.py
@@ -8274,6 +8274,82 @@ class TestBuilderHasIncompleteWork:
         assert builder._has_incomplete_work(diag) is False
 
 
+class TestBuilderHasIncompleteWorkWithoutCheckpoint:
+    """Test _has_incomplete_work detects substantive changes without a checkpoint (issue #2572)."""
+
+    def test_uncommitted_with_implementation_activity_returns_true(self) -> None:
+        """Uncommitted changes + log implementation activity = incomplete work."""
+        builder = BuilderPhase()
+        diag = {
+            "worktree_exists": True,
+            "has_uncommitted_changes": True,
+            "commits_ahead": 0,
+            "remote_branch_exists": False,
+            "pr_number": None,
+            "pr_has_review_label": False,
+            "log_has_implementation_activity": True,
+        }
+        assert builder._has_incomplete_work(diag) is True
+
+    def test_uncommitted_with_multiple_files_returns_true(self) -> None:
+        """Uncommitted changes + >=2 substantive files = incomplete work."""
+        builder = BuilderPhase()
+        diag = {
+            "worktree_exists": True,
+            "has_uncommitted_changes": True,
+            "commits_ahead": 0,
+            "remote_branch_exists": False,
+            "pr_number": None,
+            "pr_has_review_label": False,
+            "uncommitted_file_count": 3,
+        }
+        assert builder._has_incomplete_work(diag) is True
+
+    def test_uncommitted_with_one_file_no_activity_returns_false(self) -> None:
+        """Uncommitted changes + 1 file + no activity = not incomplete (could be marker)."""
+        builder = BuilderPhase()
+        diag = {
+            "worktree_exists": True,
+            "has_uncommitted_changes": True,
+            "commits_ahead": 0,
+            "remote_branch_exists": False,
+            "pr_number": None,
+            "pr_has_review_label": False,
+            "log_has_implementation_activity": False,
+            "uncommitted_file_count": 1,
+        }
+        assert builder._has_incomplete_work(diag) is False
+
+    def test_uncommitted_no_evidence_returns_false(self) -> None:
+        """Uncommitted changes + no checkpoint + no activity + 0 files = not incomplete."""
+        builder = BuilderPhase()
+        diag = {
+            "worktree_exists": True,
+            "has_uncommitted_changes": True,
+            "commits_ahead": 0,
+            "remote_branch_exists": False,
+            "pr_number": None,
+            "pr_has_review_label": False,
+            "log_has_implementation_activity": False,
+            "uncommitted_file_count": 0,
+        }
+        assert builder._has_incomplete_work(diag) is False
+
+    def test_uncommitted_with_exactly_two_files_returns_true(self) -> None:
+        """Boundary: exactly 2 uncommitted files should be treated as incomplete."""
+        builder = BuilderPhase()
+        diag = {
+            "worktree_exists": True,
+            "has_uncommitted_changes": True,
+            "commits_ahead": 0,
+            "remote_branch_exists": False,
+            "pr_number": None,
+            "pr_has_review_label": False,
+            "uncommitted_file_count": 2,
+        }
+        assert builder._has_incomplete_work(diag) is True
+
+
 class TestBuilderRecoverFromExistingWorktree:
     """Test _recover_from_existing_worktree for exit code 6/7 recovery (issue #2507)."""
 

--- a/loom-tools/tests/test_validate_phase.py
+++ b/loom-tools/tests/test_validate_phase.py
@@ -507,11 +507,12 @@ class TestValidateBuilder:
 # ---------------------------------------------------------------------------
 
 
-class TestBuilderNoAutoRecovery:
-    """Tests verifying auto-recovery was removed from builder validation.
+class TestBuilderRecoveryFromUncommittedChanges:
+    """Tests verifying auto-recovery from substantive uncommitted changes.
 
-    Note: Auto-recovery (commit, push, create PR) was removed in favor of
-    explicit failure labels. These tests verify the new behavior.
+    When the builder exits with substantive uncommitted changes in the worktree,
+    validate_builder should attempt mechanical recovery (stage, commit, push,
+    create PR) instead of immediately failing.
     """
 
     def _make_worktree(self, tmp_path: Path) -> Path:
@@ -521,69 +522,86 @@ class TestBuilderNoAutoRecovery:
         (wt / ".git").mkdir()  # Minimal git marker
         return wt
 
+    @patch("loom_tools.validate_phase._log_recovery_event")
+    @patch("loom_tools.validate_phase._report_milestone")
     @patch("loom_tools.validate_phase.subprocess.run")
     @patch("loom_tools.validate_phase._find_pr_for_issue")
     @patch("loom_tools.validate_phase._run_gh")
-    def test_worktree_with_changes_fails_instead_of_recovering(
-        self, mock_gh: MagicMock, mock_find: MagicMock, mock_run: MagicMock, tmp_path: Path
+    def test_worktree_with_changes_recovers_via_commit_push_pr(
+        self, mock_gh: MagicMock, mock_find: MagicMock, mock_run: MagicMock,
+        mock_milestone: MagicMock, mock_log_recovery: MagicMock, tmp_path: Path,
     ):
-        """Verify that uncommitted changes result in failure, not recovery."""
+        """Verify that substantive uncommitted changes trigger recovery."""
         repo = _make_repo(tmp_path)
         wt = self._make_worktree(tmp_path)
 
-        mock_gh.return_value = _completed(stdout="OPEN\n")  # issue state
+        # First _run_gh call: issue state check (OPEN)
+        # Second _run_gh call: issue title for PR
+        mock_gh.side_effect = [
+            _completed(stdout="OPEN\n"),
+            _completed(stdout="Fix the widget\n"),
+        ]
         mock_find.return_value = None  # no existing PR
 
         def side_effect_fn(*args, **kwargs):
             cmd = args[0] if args else kwargs.get("args", [])
+            cmd_str = " ".join(str(c) for c in cmd)
             # git -C worktree status --porcelain
             if "status" in cmd and "--porcelain" in cmd:
-                return _completed(stdout="M src/file.py\n")  # Has changes
-            # git -C worktree log --oneline @{upstream}..HEAD
-            if "log" in cmd and "@{upstream}" in cmd:
-                return _completed(stdout="abc123 Some commit\n")  # Has unpushed commits
-            # gh issue edit (for _mark_phase_failed)
-            if "gh" in cmd and "issue" in cmd and "edit" in cmd:
+                return _completed(stdout=" M src/file.py\n M src/other.py\n")
+            # git add
+            if "git" in cmd_str and "add" in cmd:
                 return _completed()
-            # gh issue comment (for _mark_phase_failed)
-            if "gh" in cmd and "issue" in cmd and "comment" in cmd:
+            # git commit
+            if "git" in cmd_str and "commit" in cmd:
                 return _completed()
+            # git push
+            if "git" in cmd_str and "push" in cmd:
+                return _completed()
+            # gh pr create
+            if "gh" in cmd_str and "pr" in cmd and "create" in cmd:
+                return _completed(stdout="https://github.com/org/repo/pull/99\n")
             return _completed()
 
         mock_run.side_effect = side_effect_fn
 
         result = validate_builder(42, repo, worktree=str(wt))
 
-        # Should FAIL, not RECOVER - no auto-recovery anymore
-        assert result.status == ValidationStatus.FAILED
-        assert "no auto-recovery" in result.message.lower()
+        assert result.status == ValidationStatus.RECOVERED
+        assert "commit_and_pr" == result.recovery_action
+        mock_log_recovery.assert_called_once()
 
     @patch("loom_tools.validate_phase.subprocess.run")
     @patch("loom_tools.validate_phase._find_pr_for_issue")
     @patch("loom_tools.validate_phase._run_gh")
-    def test_failure_applies_correct_label(
-        self, mock_gh: MagicMock, mock_find: MagicMock, mock_run: MagicMock, tmp_path: Path
+    def test_recovery_fails_on_git_add_marks_failed(
+        self, mock_gh: MagicMock, mock_find: MagicMock, mock_run: MagicMock, tmp_path: Path,
     ):
-        """Verify failure applies loom:failed:builder label."""
+        """Verify that failed git add results in FAILED status with diagnostics."""
         repo = _make_repo(tmp_path)
         wt = self._make_worktree(tmp_path)
 
         mock_gh.return_value = _completed(stdout="OPEN\n")
         mock_find.return_value = None
 
-        label_calls = []
-
         def side_effect_fn(*args, **kwargs):
             cmd = args[0] if args else kwargs.get("args", [])
+            cmd_str = " ".join(str(c) for c in cmd)
             if "status" in cmd and "--porcelain" in cmd:
-                return _completed(stdout="M src/file.py\n")
-            if "log" in cmd and "@{upstream}" in cmd:
-                return _completed(stdout="abc123 Commit\n")
-            if "gh" in cmd and "issue" in cmd and "edit" in cmd:
-                label_calls.append(cmd)
-                return _completed()
-            if "gh" in cmd and "issue" in cmd and "comment" in cmd:
-                return _completed()
+                return _completed(stdout=" M src/file.py\n")
+            if "git" in cmd_str and "add" in cmd:
+                return _completed(returncode=1, stderr="fatal: error")
+            # For _mark_phase_failed and _gather_builder_diagnostics
+            if "rev-parse" in cmd_str:
+                return _completed(stdout="feature/issue-42\n")
+            if "rev-list" in cmd_str:
+                return _completed(stdout="0\n")
+            if "ls-remote" in cmd_str:
+                return _completed(stdout="")
+            if "pr" in cmd_str and "list" in cmd_str:
+                return _completed(stdout="")
+            if "issue" in cmd_str and "view" in cmd_str:
+                return _completed(stdout="loom:building\n")
             return _completed()
 
         mock_run.side_effect = side_effect_fn
@@ -591,8 +609,45 @@ class TestBuilderNoAutoRecovery:
         result = validate_builder(42, repo, worktree=str(wt))
 
         assert result.status == ValidationStatus.FAILED
-        assert len(label_calls) == 1
-        assert "loom:failed:builder" in label_calls[0]
+        assert "could not stage changes" in result.message.lower()
+
+    @patch("loom_tools.validate_phase.subprocess.run")
+    @patch("loom_tools.validate_phase._find_pr_for_issue")
+    @patch("loom_tools.validate_phase._run_gh")
+    def test_only_marker_files_still_fails(
+        self, mock_gh: MagicMock, mock_find: MagicMock, mock_run: MagicMock, tmp_path: Path,
+    ):
+        """Verify that only marker/infrastructure files do not trigger recovery."""
+        repo = _make_repo(tmp_path)
+        wt = self._make_worktree(tmp_path)
+
+        mock_gh.return_value = _completed(stdout="OPEN\n")
+        mock_find.return_value = None
+
+        def side_effect_fn(*args, **kwargs):
+            cmd = args[0] if args else kwargs.get("args", [])
+            cmd_str = " ".join(str(c) for c in cmd)
+            if "status" in cmd and "--porcelain" in cmd:
+                return _completed(stdout="?? .loom-in-use\n")
+            # For _mark_phase_failed and _gather_builder_diagnostics
+            if "rev-parse" in cmd_str:
+                return _completed(stdout="feature/issue-42\n")
+            if "rev-list" in cmd_str:
+                return _completed(stdout="0\n")
+            if "ls-remote" in cmd_str:
+                return _completed(stdout="")
+            if "pr" in cmd_str and "list" in cmd_str:
+                return _completed(stdout="")
+            if "issue" in cmd_str and "view" in cmd_str:
+                return _completed(stdout="loom:building\n")
+            return _completed()
+
+        mock_run.side_effect = side_effect_fn
+
+        result = validate_builder(42, repo, worktree=str(wt))
+
+        assert result.status == ValidationStatus.FAILED
+        assert "marker files" in result.message.lower()
 
 
 # ---------------------------------------------------------------------------

